utopia-remix/app/components/projectActionContextMenu.tsx

```diff
@@ -0,0 +1,163 @@
+        )
+      },
+      [fetcher],
+    )
+
+    const menuEntries = React.useMemo((): ContextMenuEntry[] => {
+      switch (selectedCategory) {
+        case 'allProjects':
+          return [
+            {
+              text: 'Open',
+              onClick: (project) => {
+                window.open(projectEditorLink(project.proj_id), '_blank')
+              },
+            },
+            'separator',
+            {
+              text: 'Copy Link',
+              onClick: (project) => {
+                navigator.clipboard.writeText(projectEditorLink(project.proj_id))
+                // TODO notification toast
+              },
+            },
+            'separator',
+            {
+              text: 'Rename',
+              onClick: (project) => {
+                const newTitle = window.prompt('New title:', project.title)
+                if (newTitle != null) {
+                  renameProject(project.proj_id, newTitle)
+                }
+              },
+            },
+            {
+              text: 'Delete',
+              onClick: (project) => {
+                deleteProject(project.proj_id)
+              },
+            },
+          ]
+        case 'trash':
+          return [
+            {
+              text: 'Restore',
+              onClick: (project) => {
+                restoreProject(project.proj_id)
+              },
+            },
+            'separator',
+            {
+              text: 'Delete permanently',
+              onClick: (project) => {
+                destroyProject(project.proj_id)
+              },
+            },
+          ]
+        default:
+          assertNever(selectedCategory)
+      }
+    }, [selectedCategory])
+
+    return (
+      <>
+        <DropdownMenu.Portal>
+          <DropdownMenu.Content
+            style={{
+              background: 'white',
+              padding: 4,
+              boxShadow: '2px 3px 4px #dddddd',
+              border: '1px solid #ccc',
+              borderRadius: 4,
+              display: 'flex',
+              flexDirection: 'column',
+              gap: 4,
+              minWidth: 100,
+            }}
+            sideOffset={5}
+          >
+            {menuEntries.map((entry, index) => {
+              if (entry === 'separator') {
+                return (
+                  <DropdownMenu.Separator
+                    key={`separator-${index}`}
+                    style={{ backgroundColor: colors.separator, height: 1 }}
+                  />
+                )
+              }
+              return (
+                <DropdownMenu.Item
+                  key={`entry-${index}`}
+                  onClick={() => entry.onClick(project)}
+                  className={contextMenuItem()}
+                >
+                  {entry.text}
+                </DropdownMenu.Item>
+              )
+            })}
+          </DropdownMenu.Content>
+        </DropdownMenu.Portal>
+
+        <fetcher.Form />
```

Comment:
what's this self closing tag do?

Reply:
it's the form component for the fetcher used to send the requests, as per the remix docs https://remix.run/docs/en/main/hooks/use-fetcher#fetcherform